RecoHGCal/TICL/plugins/TracksterLinkingbyGNN.cc

```diff
@@ -0,0 +1,443 @@
+            auto lc = layerClusters[lc_idx];
+            vertex_energy_i.push_back(lc.energy());
+            vertex_i_x.push_back(lc.x());
+            vertex_i_y.push_back(lc.y());
+            vertex_i_z.push_back(lc.z());
+        }
+        int node_degree = ticlGraph->getNode(i).getInner().size();
+        node_degrees.push_back(node_degree);
+        
+        if (node_degree > node_degree_max){
+          node_degree_max = node_degree;
+        }
+
+        for (auto &i_neighbour : ticlGraph->getNode(i).getInner()){
+
+            const auto &ts_j = tracksters[i_neighbour];
+
+            std::vector<float> vertex_energy_j;
+            std::vector<float> vertex_j_x;
+            std::vector<float> vertex_j_y;
+            std::vector<float> vertex_j_z;
+
+            const Vector &eigenvector0_j = ts_j.eigenvectors(0);
+
+            for (auto &lc_idx : ts_j.vertices()){
+                auto lc = layerClusters[lc_idx];
+                vertex_energy_j.push_back(lc.energy());
+                vertex_j_x.push_back(lc.x());
+                vertex_j_y.push_back(lc.y());
+                vertex_j_z.push_back(lc.z());
+            }
+
+            // Create an edge between the tracksters
+            edges_src.push_back(static_cast<float>(i_neighbour));
+            edges_dst.push_back(static_cast<float>(i));
+
+            auto delta_en = std::abs(ts_i.raw_energy() - ts_j.raw_energy());
+            auto delta_z = std::abs(ts_i.barycenter().z() - ts_j.barycenter().z());
+
+            // Find energy thresholds
+            float e_thr_i = 0.01 * *std::max_element(vertex_energy_i.begin(), vertex_energy_i.end());
+            float e_thr_j = 0.01 * *std::max_element(vertex_energy_j.begin(), vertex_energy_j.end());
+
+            // Filter vertices based on energy threshold
+            std::vector<std::vector<float>> lcs_i;
+            for (int k = 0; k < (int)vertex_energy_i.size(); k++){
+                float x = vertex_i_x[k];
+                float y = vertex_i_y[k];
+                float z = vertex_i_z[k];
+                float e = vertex_energy_i[k];
+                if (e >= e_thr_i)
+                    lcs_i.push_back({x, y, z});
+            }
+
+            std::vector<std::vector<float>> lcs_j;
+            for (int k = 0; k < (int)vertex_energy_j.size(); k++){
+                float x = vertex_j_x[k];
+                float y = vertex_j_y[k];
+                float z = vertex_j_z[k];
+                float e = vertex_energy_j[k];
+                if (e >= e_thr_j)
+                    lcs_j.push_back({x, y, z});
+            }
+
+            // Calculate distances
+            auto dists = calculate_dist(lcs_i, lcs_j);
+            float min_dist = *std::min_element(dists.begin(), dists.end());
+            float max_dist = *std::max_element(dists.begin(), dists.end());
+
+            // Calculate delta_r
+            float delta_r = std::sqrt(std::pow((ts_i.barycenter().x() - ts_j.barycenter().x()), 2) +
```

Comment:
can we use delta_r^2 ? so we avoid the `std::sqrt`

Reply:
It is one of the inputs to the network and it has been trained with the distance, not squared distance...